pavuk-0.9.35/man/pavuk.1
.TH pavuk 1 "2016-12-02T00:38" "0.9.35" "Internet utils"
.SH NAME
pavuk \- HTTP, HTTP over SSL, FTP, FTP over SSL and Gopher recursive document retrieval program
.SH SYNOPSIS
.PP
.B pavuk [-mode {normal | resumeregets | singlepage | singlereget | sync | dontstore | ftpdir | mirror}]
.I [-X] [-runX] [-bg/-nobg] [prefs/-noprefs] [-h] [-v] [-progress/-noprogress] [-stime/-nostime] [-xmaxlog $nr]
.I [-logfile $file] [-slogfile $file] [-auth_file $file] [-msgcat $dir] [-language $str] [-gui_font $font] [-quiet/-verbose
.I [-read_css/-noread_css]
.I [-cdir $dir]
.I [-scndir $dir]
.I [-scenario $str] [-dumpscn $filename]
.I [-lmax $nr] [-dmax $nr]
.I [-leave_level $nr]
.I [-maxsize $nr] [-minsize $nr]
.I [-asite $list] [-dsite $list]
.I [-adomain $list] [-ddomain $list]
.I [-asfx $list] [-dsfx $list]
.I [-aprefix $list] [-dprefix $list]
.I [-amimt $list] [-dmimet $list]
.I [-pattern $pattern] [-url_pattern $pattern]
.I [-rpattern $regexp] [-url_rpattern $regexp]
.I [-skip_pattern $pattern] [-skip_url_pattern $pattern]
.I [-skip_rpattern $regexp] [-skip_url_rpattern $regexp]
.I [-newer_than $time] [-older_than $time]
.I [-schedule $time] [-reschedule $nr]
.I [-dont_leave_site/-leave_site] [-dont_leave_dir/-leave_dir]
.I [-http_proxy $site[:$port]] [-ftp_proxy $site[:$port]]
.I [-ssl_proxy $site[:$port]] [-gopher_proxy $site[:$port]]
.I [-ftp_httpgw/-noftp_httpgw] [-ftp_dirtyproxy/-noftp_dirtyproxy]
.I [-gopher_httpgw/-nogopher_httpgw]
.I [-noFTP/-FTP] [-noHTTP/-HTTP] [-noSSL/-SSL] [-noGopher/-Gopher]
.I [-FTPdir/-noFTPdir]
.I [-noCGI/-CGI]
.I [-FTPlist/-noFTPlist] [-FTPhtml/-noFTPhtml]
.I [-noRelocate/-Relocate] [-force_reget/-noforce_reget]
.I [-nocache/-cache] [-check_size/-nocheck_size]
.I [-noRobots/-Robots] [-noEnc/-Enc]
.I [-auth_name $user] [-auth_passwd $pass] [-auth_scheme 1/2/3/4/user/Basic/Digest/NTLM]
.I [-auth_reuse_nonce/-no_auth_reuse_nonce]
.I [-http_proxy_user $user] [-http_proxy_pass $pass] [-http_proxy_auth 1/2/3/4/user/Basic/Digest/NTLM]
.I [-auth_reuse_proxy_nonce/-no_auth_reuse_proxy_nonce]
.I [-ssl_key_file $file] [-ssl_cert_file $file] [-ssl_cert_passwd $pass]
.I [-from $email] [-send_from/-nosend_from] [-identity $str] [-auto_referer/-noauto_referer]
.I [-referer/-noreferer]
.I [-alang $list] [-acharset $list]
.I [-retry $nr] [-nregets $nr] [-nredirs $nr] [-rollback $nr] [-sleep $nr] [-timeout $nr]
.I [-preserve_time/-nopreserve_time] [-preserve_perm/-nopreserve_perm] [-preserve_slinks/-nopreserve_slinks]
.I [-bufsize $nr] [-maxrate $nr] [-minrate $nr]
.I [-user_condition $str]
.I [-cookie_file $file] [-cookie_send/-nocookie_send] [-cookie_recv/-nocookie_recv] [-cookie_update/-nocookie_update] [-cookies_max $nr] [-disabled_cookie_domains $list]
.I [-disable_html_tag $TAG,[$ATTRIB][;...]] [-enable_html_tag $TAG,[$ATTRIB][;...]]
.I [-tr_del_chr $str] [-tr_str_str $str1 $str2] [-tr_chr_chr $chrset1 $chrset2]
.I [-index_name $str] [-store_index/-nostore_index]
.I [-store_name $str] [-debug/-nodebug] [-debug_level $level]
.I [-browser $str]
.I [-urls_file $file]
.I [-file_quota $nr] [-trans_quota $nr] [-fs_quota $nr] [-enable_js/-disable_js]
.I [-fnrules $t $m $r] [-store_info/-nostore_info]
.I [-all_to_local/-noall_to_local] [-sel_to_local/-nosel_to_local]
.I [-all_to_remote/-noall_to_remote]
.I [-url_strategie $strategie]
.I [-remove_adv/-noremove_adv] [-adv_re $RE]
.I [-check_bg/-nocheck_bg]
.I [-send_if_range/-nosend_if_range]
.I [-sched_cmd $str]
.I [-unique_log/-nounique_log]
.I [-post_cmd $str] [-ssl_version $v] [-unique_sslid/-nounique_sslid]
.I [-aip_pattern $re] [-dip_pattern $re]
.I [-use_http11/-nouse_http11]
.I [-local_ip $addr]
.I [-request $req]
.I [-formdata $req] [-httpad $str]
.I [-nthreads $nr] [-immesg/-noimmesg]
.I [-dumpfd $nr] [-dump_urlfd $nr]
.I [-unique_name/-nounique_name]
.I [-leave_site_enter_dir/-dont_leave_site_enter_dir]
.I [-max_time $nr]
.I [-del_after/-nodel_after] [-singlepage/-nosinglepage]
.I [-dump_after/-nodump_after] [-dump_response/-nodump_response]
.I [-auth_ntlm_domain $str] [-auth_proxy_ntlm_domain $str]
.I [-js_pattern $re] [-follow_cmd $str]
.I [-retrieve_symlink/-noretrieve_symlink]
.I [-js_transform $p $t $h $a] [-js_transform2 $p $t $h $a]
.I [-ftp_proxy_user $str] [-ftp_proxy_pass $str]
.I [-limit_inlines/-dont_limit_inlines]
.I [-ftp_list_options $str] [-fix_wuftpd_list/-nofix_wuftpd_list]
.I [-post_update/-nopost_update] [-info_dir $dir]
.I [-mozcache_dir $dir] [-aport $list] [-dport $list]
.I [-hack_add_index/-nohack_add_index]
.I [-default_prefix $str] [-rsleep/-norsleep]
.I [-ftp_login_handshake $host $handshake]
.I [-js_script_file $file]
.I [-dont_touch_url_pattern $pat]
.I [-dont_touch_url_rpattern $pat]
.I [-dont_touch_tag_rpattern $pat]
.I [-tag_pattern $tag $attrib $url]
.I [-tag_rpattern $tag $attrib $url]
.I [-nss_cert_dir $dir]
.I [-nss_accept_unknown_cert/-nonss_accept_unknown_cert]
.I [-nss_domestic_policy/-nss_export_policy]
.I [-[no]verify]
.I [-tlogfile $file]
.I [-trelative {object | program}]
.I [-transparent_proxy FQDN[:port]]
.I [-transparent_ssl_proxy FQDN[:port]]
.I [-sdemo]
.I [-noencode]
.I [URLs]
.br
.PP
.B pavuk -mode {normal | singlepage | singlereget}
.I [-base_level $nr]
.br
.PP
.B pavuk -mode sync
.I [-ddays $nr] [-subdir $dir] [-remove_old/-noremove_old]
.br
.PP
.B pavuk -mode resumeregets
.I [-subdir $dir]
.br
.PP
.B pavuk -mode linkupdate
.I [-X] [-h] [-v] [-cdir $dir] [-subdir $dir] [-scndir $dir] [-scenario $str]
.PP
.B pavuk -mode reminder
.I [-remind_cmd $str]
.br
.PP
.B pavuk -mode mirror
.I [-subdir $dir] [-remove_old/-noremove_old] [-remove_before_store/-noremove_before_store] [-always_mdtm/-noalways_mdtm]
.br
.PP
.SH DESCRIPTION
.sp
This manual page describes how to use pavuk. Pavuk can be used to
mirror contents of internet/intranet servers and to maintain copies in
a local tree of documents.  Pavuk stores retrieved documents in
locally mapped disk space. The structure of the local tree is the same
as the one on the remote server. Each supported service (protocol) has
its own subdirectory in the local tree.  Each referenced server has
its own subdirectory in these protocols subdirectories; followed by
the port number on which the service resides, delimited by character
'_'. Then the directory structure follows.  As of version 0.9pl19 this
can be be changed. With the option
.B -fnrules
you can change the default layout of the local document tree,
without losing link consistency.
.br
With
.B pavuk
it is possible to have up-to-date copies of remote documents in the
local disk space.
.br
As of version 0.3pl2, pavuk can automatically restart broken connections, and reget
partial content from an FTP server (which must support the
.B REST
command), from a properly configured HTTP/1.1 server, or from a HTTP/1.0
server which supports
.B Ranges.
.br
As of version 0.6 it is possible to handle configurations via so
called scenarios.  The best way to create such a configuration file is
to use the X Window interface and simply save the created
configuration. The other way is to use the -dumpscn switch.
.br
As of version 0.7pl1 it is possible to store authentification information
into an authinfo file, which pavuk can then parse and use.
.br
As of version 0.8pl4 pavuk can fetch documents for use in a local
proxy/cache server without storing them to local documents tree.
.br
As of version 0.9pl4 pavuk supports
.B SOCKS (4/5)
proxies if you have the required libraries.
.br
As of version 0.9pl12 pavuk can preserve permissions of remote files and
symbolic links, so it can be used for powerful FTP mirroring.
.br
Pavuk supports SSL connections to FTP servers, if you specify ftps://
URL instead of ftp://.
.br
Pavuk can automatically handle file names with unsafe characters for filesystem.
This is yet implemented only for Win32 platform and it is hard coded.
.br
Pavuk can now use \fBHTTP/1.1\fR protocol for communication with HTTP servers.
It can use persistent connections, so one TCP connection should be used to
transfer several documents without closing it. This feature saves network
bandwidth and also speedup network communication.
.br
Pavuk can do configurable \fBPOST\fR requests to HTTP servers and support
also file uploading via HTTP POST request.
.br
Pavuk can automatically fill found HTML forms, if user will supply data
for its fields before with option \fB-formdata\fR.
.br
Pavuk can run configurable number of concurrently running downloading threads when
compiled with multithreading support.

.SH Format of supported URLs
.sp
.sp
\fIHTTP\fR
.br
http://[[user][:password]@]host[:port][/document]
.br
[[user][:password]@]host[:port][/document]
.sp
\fIHTTPS\fR
.br
https://[[user][:password]@]host[:port][/document]
.br
ssl[.domain][:port][/document]
.sp
\fIFTP\fR
.br
ftp://[[user][:password]@]host[:port][/relative_path][;type=x]
.br
ftp://[[user][:password]@]host[:port][//absolute_path][;type=x]
.br
ftp[.domain][:port][/document][;type=x]
.sp
\fIFTPS\fR
.br
ftps://[[user][:password]@]host[:port][/relative_path][;type=x]
.br
ftps://[[user][:password]@]host[:port][//absolute_path][;type=x]
.br
ftps[.domain][:port][/document][;type=x]
.sp
\fIGopher\fR
.br
gopher://host[:port][/type[document]]
.br
gopher[.domain][:port][/type[document]]

.SH Default mapping of URLs to local filenames
.sp
.sp
\fIHTTP\fR
.br
http://[[user][:password]@]host[:port][/document][?query]
.br
to
.br
http/host_port/[document][?query]
.sp
\fIHTTPS\fR
.br
https://[[user][:password]@]host[:port][/document][?query]
.br
to
.br
https/host_port/[document][?query]
.sp
\fIFTP\fR
.br
ftp://[[user][:password]@]host[:port][/path]
.br
to
.br
ftp/host_port/[path]
.sp
\fIFTPS\fR
.br
ftps://[[user][:password]@]host[:port][/path]
.br
to
.br
ftps/host_port/[path]
.sp
\fIGopher\fR
.br
gopher://host[:port][/type[document]]
.br
to
.br
gopher/host_port/[type[document]]

\fINOTE:\fR Pavuk will use the string with which it queries the target
server as the name of the results file. This file name may, in some cases,
contain punctuations such as \fB$,?,=,&\fR etc. Such punctuation can cause
problems when you are trying to browse downloaded files with your browser or
you are trying to process downloaded files with shell scripts or view files
with file management utilities which reference the name of the results file.
If you believe that this maybe causing problems for you, then you can remove
all punctuation from the result file name with the option:
\fB-tr_del_chr [:punct:]\fR or with other options for adjusting filenames.


.SH OPTIONS
.sp
 All options are case insensitive.

.SH List of options chapters
.sp
.I Mode
.br
.I Help
.br
.I Indicate/Logging/Interface options
.br
.I Netli options
.br
.I Special start
.br
.I Scenario/Task options
.br
.I Directory options
.br
.I Preserve options
.br
.I Proxy options
.br
.I Proxy Authentification
.br
.I Protocol/Download Option
.br
.I Authentification
.br
.I Site/Domain/Port Limitation Options
.br
.I Limitation Document properties
.br
.I Limitation Document name
.br
.I Limitation Protocol Option
.br
.I Other Limitation Options
.br
.I Javascript support
.br
.I Cookie
.br
.I HTML rewriting engine tuning options
.br
.I Filename/URL Conversion Option
.br
.I Other Options
.br

.SH Mode
.sp
.TP
.I -mode {normal, linkupdate, sync, singlepage, singlereget, resumeregets}
Set operation mode.
.br
.B normal
- retrieves recursive documents
.br
.B linkupdate
- update remote URLs in local HTML documents to local URLs if these URLs
exist in the local tree
.br
.B sync
- synchronize remote documents with local tree (if a local copy of
a document is older than remote, the document is retrieved again,
otherwise nothing happens)
.br
.B singlepage
- URL is retrieved as one page with all inline objects (picture, sound ...)
this mode is now obsoleted by \fB-singlepage\fR option.
.br
.B resumeregets
- pavuk scans the local tree for files that were not retrieved fully
and retrieves them again (uses partial get if possible)
.br
.B singlereget
- get URL until it is retrieved in full
.br
.B dontstore
- transfer page from server, but don't store it to the local tree.
This mode is suitable for fetching pages that are held in a local
proxy/cache server.
.br
.B reminder
- used to inform the user about changed documents
.br
.B ftpdir
- used to list of contents of FTP directories
.br
.sp
default operation mode is
.B normal
mode.

.SH Help
.sp
.TP
.I -h
Print long verbose help message
.TP
.I -v
Show version informations and configuration at compilation time.

.SH Indicate/Logging/Interface options
.sp
.TP
.I -quiet
Don't show any messages on the screen.
.TP
.I -verbose
Force to show output messages on the screen (default)
.TP
.I -progress/-noprogress
Show retrieving progress while running in the terminal (default is progress off)
.TP
.I -stime/-nostime
Show start and end time of transfer. (default isn't this information shown)
.TP
.I -xmaxlog $nr
Maximum number of log lines in the Log widget. 0 means unlimited.
This option is available only when compiled with the GTK+ GUI. (default value is 0)
.TP
.I -logfile $file
File where all produced messages are stored.
.TP
.I -unique_log/-nounique_log
When logfile as specified with the option
.B -logfile
is already used by another process, try to generate new unique name
for the log file. (default is this option turned off)
.TP
.I -slogfile $file
File to store short logs in. This file contains one line of
informations per processed document.  This is meant to be used in
connection with any sort of script to produce some statistics, for
validating links on your website, or for generating simple sitemaps.
Multiple pavuk processes can use this file concurrently, without
overwriting each others entries.
Record structure:
.br
.sp
.RS
.nf
- \fBPID\fR of pavuk process
- \fBTIME\fR current time
- \fBCOUNTER\fR in the format current/total number of URLs
- \fBSTATUS\fR contains the type of the error: FATAL, ERR,
  WARN or OK
- \fBERRCODE\fR is the number code of the error
  (see errcode.h in pavuk sources)
- \fBURL\fR of the document
- \fBPARENTURL\fR first parent document of this URL
  (when it doesn't have parent - [none])
- \fBFILENAME\fR is the name of the local file the
  document is saved under
- \fBSIZE\fR size of requested document if known
- \fBDOWNLOAD_TIME\fR time which takes downloading of this
  document in format seconds.mili_seconds
- \fBHTTPRESP\fR contains the first line of the HTTP server
  response
.fi
.RE
.TP
.I -language $str
Native language that pavuk should use for communication with its

pavuk-0.9.35/NEWS
This file contains new features introduced by particular pavuk versions.
For more precise description of changes look into ChangeLog file.
Fileds signed as <CLI> are for commandline new features, <GUI> are for new
feature available only for GUI, <DOC> significant changes in documentation.

version 0.9pl22
---------------
<CLI> new option -ssl_version used to set required SSL protocol version
<CLI> new options -unique_sslid/-nounique_sslid used to specify if pavuk
      should use unique SSL ID with each SSL session
<GUI> New Append URL dialog used to append URLs while running downloading
      process
<CLI> new option -httpad used for adding user defined HTTP headers to HTTP
      requests
<CLI> new option -statfile for saving download progress statistics to file
      when finished
<GUI> new dialog window for previewing statistics reports
<CLI> new option -debug_level to enable control on amount of debug messages
<CLI> new WIN32 specific option -ewait to disable pavuk console close when
      pavuk finished his job
<GUI> new posibility to save tree structure from URL tree preview dialog

version 0.9pl23
---------------
<CLI> new option -aip_pattern & -dip_pattern to specify allowed/disallowed
      server IP addreses with regular expressions
<CLI> new option -site_level for limiting through how many site in URL tree
      can pavuk go
<CLI> new mode "ftpdir" for listing FTP directories
<CLI> new option -site_level for limiting how many site levels can be
      downloaded
<CLI> new protocol added - FTP over SSL
<CLI> new option -FTPS/-noFTPS for enabling/disabling of documents via
      FTP over SSL
<CLI> new option -ssl_cipher_list to specify list of preffered ciphers in 
      SSL communication
<CLI> new protocol added - HTTP/1.1
<CLI> new option -use_http11/-nouse_http11 for enabling/disabling use of 
      HTTP/1.1 protocol in communication with HTTP server
<CLI> new option -max_time to specify maximal time of downloading
<CLI> new option -local_ip to binf specified network interface on multihomed
      hosts

version 0.9pl24
---------------
<CLI> new option -request for specify richer information for request, this
      option is required for doing HTTP POST requests
<CLI> new option -hash_size used to set size of internal hash tables for
      URL and filename lookup. Usable for performance tuning when mirrorng
      huge amount of URLs
<CLI> extended posibilities of option -fnrules to be able to perform several
      functions for generating local names
<GUI> implemented droping of URLs into URL Append dialog
<GUI> implemented posibility to watch downloading process inside 
      URL tree preview dialog

version 0.9pl25
---------------
<GUI> removed support for Xt GUI
<CLI> posibility to use multiple HTTP proxy servers with round robin scheduling
<CLI/GUI> optional multithreading support (see option -nthreads)
<CLI> new multithreading related pair of options -immesg/-noimmesg for
      controling per thread bufering of messages till download is finished
<CLI> posibility to fill noninteractively matching HTML forms (see option
      -formdata in manual)
<CLI> new option -dumpfd to be able directly dump documents to pipe instead 
      of file in document tree
<GUI> 'Load scenario' now resets configuration before loading new scenario
<GUI> Added new menu entry to add scenario to current setup

version 0.9pl26
---------------
<CLI> added new option -del_after to allow removing of remote FTP documents
      after successfull download
<CLI> added new options -unique_name/-nounique_name which allows to to turn
      on/off generating of unique names for documents using numbering of 
      overlaying documents
<CLI> it is now possible to specify in -request option to specify also
      filename in which will be specified starting document stored
<CLI> added new option -dump_urlsfd to enable outputing URLs from downloaded
      HTML documents to selected file descriptor - usable for scripting
<DOC> added new document file wget-pavuk.HOWTO to aid wget users start using
      pavuk
<CLI> new options -singlepage/-nosinglepage to allow using single page transfers
      in different modes (this makes -mode singlepage obsoleted)

version 0.9pl27
---------------
<GUI> added support to retieve URL from currently opened document in Netscape
      browser window
<CLI> options  -disable_html_tag / -enable_html_tag now accept parameter
      "all" to disable / enable all HTML tags
<CLI> added support for reading MSIE cache on WIN32. Use options -ie_cache
      and -noie_cache to enable/disable this feature
<CLI> new macro for -fnrules option %q for handling query strings from POST
      request
<CLI> added new functions for -fnrules option - "sif", "!", "|", "&" .
<CLI> added support for GNU getopt() like long/short commandline options
      You can now use this syntax of options:
	-lmax 3
	--lmax 3
	--lmax=3
	-l 3
	-l3
      It is also able to combine multiple short options.
      For example -xl3 means -X -lmax 3
<CLI> added new options -dump_after/-nodump_after. This option can be used
      with option -dumpfd x , and you can control with it when document will
      writen to output. When -dump_after is active, document will be dumped
      just after download and processing of HTML documents. Else document
      will be dumped during download. With this option you can also use
      option -dumpfd succesfully in multithreaded version of pavuk without
      mixing documents on output
<CLI> added support for processing simple JS patterns in DOM event tags and
      script bodies and javascript:... URLs
<CLI> added support for NTLM authorization
<CLI> added support for HTTP/1.0 persistent proxy connections
<CLI> added new option -js_pattern which will allow you to specify custom
      JS RE patterns for processing
<CLI> added option -follow_cmd, which allows to cotrol pavuk with external
      command. By exit code you can tell if pavuk can follow links from
      current doc.

pavuk-0.9.35/src/pavuk.c
/***************************************************************************/
/*    This code is part of WWW grabber called pavuk                        */
/*    Copyright (c) 1997 - 2001 Stefan Ondrejicka                          */
/*    Distributed under GPL 2 or later                                     */
/***************************************************************************/

#include "config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <netdb.h>
#include <errno.h>
#include <unistd.h>
#include <time.h>
#include <limits.h>
#include <pwd.h>
#include <signal.h>

#ifdef __CYGWIN__
#include <windows.h>
#endif

#include "recurse.h"
#include "http.h"
#include "ftp.h"
#include "update_links.h"
#include "mode.h"
#include "url.h"
#include "dns.h"
#include "ainterface.h"
#include "times.h"
#include "schedule.h"
#include "dlhash_tools.h"
#include "nscache.h"
#include "log.h"
#include "authinfo.h"
#include "cookie.h"
#include "net.h"
#include "gui_api.h"
#include "gui.h"
#include "myssl.h"

#define STDOUT  1

_config_struct_t cfg;

#ifdef WIN32
/*
 * read string value from HKEY_LOCAL_MACHINE
 */
char *read_lmachine_registry_val(char *path, char *var)
{
  HKEY hKey;
  char rpath[2048];
  DWORD sz = sizeof(rpath);
  DWORD type;
  char *rv = NULL;

  if(RegOpenKeyEx(HKEY_LOCAL_MACHINE, path, 0, KEY_READ,
      &hKey) != ERROR_SUCCESS)
    return NULL;

  if(RegQueryValueEx(hKey, var, NULL, &type, rpath, &sz) == ERROR_SUCCESS)
    rv = tl_strdup(rpath);

  RegCloseKey(hKey);
  return rv;
}

#endif

static char *pavuk_get_install_path(void)
{
  char *rv = NULL;
#ifdef WIN32
  rv = read_lmachine_registry_val("Software\\Stefan Ondrejicka\\Pavuk",
    "Install Path");

  if(rv)
  {
    char *p = rv;
    rv = cvt_win32_to_unix_path(p);
    _free(p);
  }
  else
    rv = tl_strdup("/cygdrive/c");
#else
#ifdef INSTALL_PREFIX
  rv = tl_strdup(INSTALL_PREFIX);
#else
  rv = tl_strdup("/usr/local");
#endif
#endif

  return rv;
}

static void pavuk_quit(int signum)
{
  printf(gettext("QUIT signal catched\n"));
  cfg.xi_face = FALSE;
  cfg.rbreak = TRUE;
  cfg.stop = TRUE;
}

static void pavuk_alarm(int signum)
{
  printf(gettext("Program has been forcefully terminated\n"));
  exit(10);
}

static void pavuk_end(int signum)
{
  signal(SIGALRM, pavuk_alarm);
  signal(SIGINT, SIG_IGN);
  signal(SIGTERM, SIG_IGN);
  printf(gettext("TERM signal catched\n"));
  cfg.xi_face = FALSE;
  cfg.rbreak = TRUE;
  cfg.stop = TRUE;
  alarm(1);
}

static void init_values(int argc, char **argv)
{
  char *d;
  uid_t uid;
  time_t __time = time(NULL);
  char pom[PATH_MAX];
  struct hostent *hp = NULL;
  struct passwd *pwstruct;
#if defined(GETTEXT_NLS) || defined(I_FACE)
  int i;
#endif
  memset(&cfg, '\0', sizeof(cfg));
#ifdef I_FACE
  memset(&gui_cfg, '\0', sizeof(gui_cfg));
#endif

  cfg.prg_path = argv[0];
  cfg.install_path = pavuk_get_install_path();

#ifdef HAVE_MT
  mt_init();
#endif

  init_locale_env();

#ifdef GTK_FACE
  cfg.path_to_home = tl_strdup(g_get_home_dir());
#endif
  if(!cfg.path_to_home)
    cfg.path_to_home = tl_strdup(getenv("HOME"));
  if(!cfg.path_to_home)
    cfg.path_to_home = tl_strdup("/tmp/");

#ifdef HAVE_TZSET
  tzset();
#endif

#ifdef I_FACE
  cfg.done = FALSE;
#endif
  cfg.urlstack = NULL;
  cfg.urls_in_dir = NULL;
  cfg.total_cnt = 0;
  cfg.time = new_tm(localtime(&__time));
  cfg.time->tm_year += 1900;
  cfg.fail_cnt = 0;
  cfg.docnr = 0;

  cfg.url_hash_tbl = NULL;
  cfg.fn_hash_tbl = NULL;
  cfg.last_used_proxy_node = NULL;

  cfg_setup_default();

#ifdef GETTEXT_NLS
#ifdef GETTEXT_DEFAULT_CATALOG_DIR
  cfg.msgcatd = tl_strdup(GETTEXT_DEFAULT_CATALOG_DIR);
#endif
#ifdef __CYGWIN__
  _free(cfg.msgcatd);
  cfg.msgcatd = tl_str_concat(NULL, cfg.install_path, "/share/locale", NULL);
#endif
#else
  cfg.language = "C";
#endif

  _INIT_NLS;

#ifdef SOCKS
  SOCKSinit(argv[0]);
#endif

  if(!(d = getenv("USER")))
  {
    uid = getuid();
    if((pwstruct = getpwuid(uid)))
      d = tl_strdup(pwstruct->pw_name);
  }
  else
    d = tl_strdup(d);

  if(gethostname(pom, sizeof(pom)))
  {
    perror("gethostname");
  }
  else
  {
    cfg.local_host = tl_strdup(pom);
    hp = gethostbyname(pom);
  }

  if(hp)
  {
    if(d)
      snprintf(pom, sizeof(pom), "%s@%s", d, hp->h_name);
    else
      snprintf(pom, sizeof(pom), "pavuk@%s", hp->h_name);
  }
  else
  {
    if(d)
      snprintf(pom, sizeof(pom), "%s@unknown.sk", d);
    else
      strcpy(pom, "pavuk@unknown.sk");
  }

  _free(d);
  cfg.from = tl_strdup(pom);

  cfg_load_setup();

  cfg.xi_face = FALSE;

#ifdef I_FACE
  cfg.stop = FALSE;
  cfg.rbreak = FALSE;
  for(i = 1; i < argc; i++)
  {
    /*** load preferences ***/
    if(!strcasecmp(argv[i], "-prefs"))
      cfg.use_prefs = TRUE;
    else if(!strcasecmp(argv[i], "-noprefs"))
      cfg.use_prefs = FALSE;
    /**** we want to run GUI ****/
    else if(!strcasecmp(argv[i], "-X"))
      cfg.xi_face = TRUE;
  }
  if(cfg.use_prefs && cfg.xi_face)
    cfg_load_pref();
#endif

  _INIT_NLS;

#ifdef GETTEXT_NLS
/*** these parameters have to be resolved before each other ***/
  for(i = 1; i < argc; i++)
  {
    if(!strcasecmp(argv[i], "-msgcat"))
    {
      i++;
      if(i < argc)
      {
        cfg.msgcatd = tl_strdup(argv[i]);
      }
      else
      {
        xprintf(0, gettext("Not enough number of parameters \"-msgcat\"\n"));
        usage();
      }

    }
    if(!strcasecmp(argv[i], "-language"))
    {
      i++;
      if(i < argc)
      {
        cfg.language = tl_strdup(argv[i]);
      }
      else
      {
        xprintf(0,
          gettext("Not enough number of parameters \"-scenario\"\n")); /* FIXME: Wrong string? */
        usage();
      }
    }
  }
#endif

  _INIT_NLS;

  cfg_setup_cmdln(argc, argv);

  /**** if requested, create GUI ****/
  if(cfg.xi_face)
  {
#if defined I_FACE && !defined HAVE_MT
    dns_serv_start();
#endif
    gui_start(&argc, argv);
  }


#ifdef GETTEXT_NLS
  cfg.language = tl_strdup(getenv("LC_MESSAGES"));
#endif

  if(cfg.dumpfd >= 0)
  {
    if((fcntl(cfg.dumpfd, F_GETFD) < 0) && (errno == EBADF))
    {
      xprintf(0,
        gettext("Error: Supplied bad file descriptor in -dumpfd option\n"),
        cfg.dumpfd);
      exit(PAVUK_EXIT_CFG_ERR);
    }
  }

  if(cfg.dump_urlfd >= 0)
  {
    if((fcntl(cfg.dump_urlfd, F_GETFD) < 0) && (errno == EBADF))
    {
      xprintf(0,
        gettext
        ("Error: Supplied bad file descriptor in -dump_urlfd option\n"),
        cfg.dumpfd);
      exit(PAVUK_EXIT_CFG_ERR);
    }
  }

  cfg.url_hash_tbl =
    dlhash_new(cfg.hash_size, url_key_func, url_hash_func,
    dllist_url_compare);
  dlhash_set_free_func(cfg.url_hash_tbl, url_free_func, NULL);
  cfg.fn_hash_tbl =
    dlhash_new(cfg.hash_size, fn_key_func, str_hash_func, str_comp_func);
  dlhash_set_free_func(cfg.fn_hash_tbl, NULL, NULL);

  memset(&cfg.local_ip_addr, '\0', sizeof(cfg.local_ip_addr));
  if(cfg.local_ip && net_host_to_in_addr(cfg.local_ip, &cfg.local_ip_addr))
  {
    xherror(cfg.local_ip);
  }

  if(cfg.cache_dir)
  {
    d = cfg.cache_dir;
    cfg.cache_dir = get_abs_file_path_oss(cfg.cache_dir);
    free(d);
  }
  else
  {
    getcwd(pom, sizeof(pom));
    cfg.cache_dir = tl_strdup(pom);
  }

  if(cfg.subdir)
  {
    d = cfg.subdir;
    cfg.subdir = get_abs_file_path_oss(cfg.subdir);
    free(d);
    if(tl_is_dirname(cfg.subdir))
      *(cfg.subdir + strlen(cfg.subdir) - 1) = '\0';
  }

  cfg.prev_mode = cfg.mode;

  if(cfg.save_scn)
  {
    if(cfg.scndir)
    {
      snprintf(pom, sizeof(pom), "%s/%s", cfg.scndir, cfg.save_scn);
      cfg_dump(pom);
    }
    else
    {
      xprintf(0,
        gettext
        ("WARNING: scndir not specified - saving to current directory\n"));
      cfg_dump(cfg.save_scn);
    }
    exit(PAVUK_EXIT_OK);
  }

  if(cfg.cookie_file)
    cookie_read_file(cfg.cookie_file);

  if(cfg.auth_file)
    authinfo_load(cfg.auth_file);

  log_start(cfg.logfile);

  if(!cfg.sched_cmd)
    cfg.sched_cmd = tl_strdup(AT_CMD);

  if(cfg.schtime)
  {
    _free(cfg.time);
    cfg.time = new_tm(localtime(&cfg.schtime));
  }

  if(!cfg.index_name)
    cfg.index_name = tl_strdup("_._.html");

#ifdef USE_SSL
  my_ssl_init_once();
#endif

  if(cfg.bgmode)
  {
    pid_t ppid;

    ppid = fork();
    if(ppid < 0)
    {
      xperror("fork");
      xprintf(1,
        gettext
        ("Unable to fork pavuk to background - running in foreground\n"));
    }
    else if(ppid != 0)
    {
      xprintf(0, gettext("Pavuk will run at backround as PID %d\n"),
        (int) ppid);
      exit(PAVUK_EXIT_CFG_ERR);
    }
#ifdef __CYGWIN__
    FreeConsole();
#endif
  }
}

static void read_urls(char *filename)
{
  bufio *fd;
  char lnbuf[4096];
  int n;
  bool_t isstdin;


  isstdin = !strcmp(cfg.urls_file, "-");

  DEBUG_MISC(gettext("reading URLs from file - %s\n"), filename);
  if(isstdin)
    fd = bufio_fdopen(0);
  else
    fd = bufio_open(filename, O_BINARY | O_RDONLY);

  if(!fd)
  {
    xperror(filename);
    return;
  }

  while((n = bufio_readln(fd, lnbuf, sizeof(lnbuf))) > 0)
  {
    strip_nl(lnbuf);
    if(!strcmp(lnbuf, "."))
      break;

    if(lnbuf[0])
    {
      url_info *ui;

      ui = url_info_new(lnbuf);
      cfg.request = dllist_append(cfg.request, (dllist_t) ui);
    }
  }

  if(n < 0)
    xperror("reading stdin");

  if(isstdin)
    bufio_free(fd);
  else
    bufio_close(fd);
}

int main(int argc, char **argv)
{
  time_t __time = time(NULL);

  init_values(argc, argv);
  atexit(pavuk_do_at_exit);

  /*
     pro: We do not fully trust pavuk to stop working after the timeout
     expired, so we order the OS to send us an ALARM signal one minute
     after the timeout expires. If the program still runs at that time
     it will be aborted once the signal strikes home.
   */
  if(cfg.max_time > 0)
  {
    alarm((int) ((cfg.max_time + 1) * 60));
  }

  if(cfg.urls_file)
  {
    read_urls(cfg.urls_file);
    _free(cfg.urls_file);
  }

  /*
     pro: Set seed for random generator; needed to find a port for
     active ftp.
   */
  srand(time(NULL) ^ getpid());

  signal(SIGINT, pavuk_end);
  signal(SIGTERM, pavuk_end);
  signal(SIGALRM, pavuk_end);
  signal(SIGPIPE, SIG_IGN);
/**** spustenie algoritmu alebo rozhrania ****/
/**** FIXME: Translate me!                ****/
  if(cfg.xi_face)
  {
    cfg.prev_mode = cfg.mode;
    cfg.mode_started = FALSE;

    gui_main();
  }
  else
  {
    signal(SIGQUIT, pavuk_quit);

    if(cfg.schtime)
    {
      cfg.schtime = (time_t) 0;
      if(at_schedule())
      {
        xprintf(0, gettext("Error scheduling\n"));
      }
    }
    else
    {
      if(cfg.reschedh)
      {
        __time += 3600 * cfg.reschedh;
        _free(cfg.time);
        cfg.time = new_tm(localtime(&__time));
        at_schedule();
      }
      absi_restart();
    }
  }

  log_start(NULL);

  print_all_url_infos();
  print_all_dns_infos();
  
  return cfg.fail_cnt ? PAVUK_EXIT_DOC_ERR : PAVUK_EXIT_OK;
}
